enterprise/suppress-user/internal/badgerdb/badgerdb.go

```diff
@@ -0,0 +1,307 @@
+				err = wb.Delete([]byte(key))
+			} else {
+				err = wb.Set([]byte(key), []byte(""))
+			}
+			if err != nil {
+				return fmt.Errorf("could not add key %s (canceled:%t) in write batch: %w", key, suppression.Canceled, err)
+			}
+		}
+
+	}
+	if err := wb.Set([]byte(tokenKey), token); err != nil {
+		return fmt.Errorf("could not add token key %s in write batch: %w", tokenKey, err)
+	}
+	if err := wb.Flush(); err != nil {
+		return fmt.Errorf("could not flush write batch: %w", err)
+	}
+	return nil
+}
+
+// start the repository
+func (b *Repository) start() error {
+	b.closed = make(chan struct{})
+	var seeder io.Reader
+	if _, err := os.Stat(b.path); os.IsNotExist(err) && b.seederSource != nil {
+		if seeder, err = b.seederSource(); err != nil {
+			return err
+		}
+	}
+
+	opts := badger.
+		DefaultOptions(b.path).
+		WithLogger(blogger{b.log}).
+		WithCompression(options.None).
+		WithIndexCacheSize(16 << 20). // 16mb
+		WithNumGoroutines(b.maxGoroutines)
+	var err error
+	b.db, err = badger.Open(opts)
+	if err != nil {
+		return err
+	}
+
+	if seeder != nil {
+		restoreDone := lo.Async(func() error {
+			if err := b.Restore(seeder); err != nil {
+				b.log.Error("Failed to restore badgerdb", "error", err)
+				return err
+			}
+			return nil
+		})
+		select {
+		case <-restoreDone:
+		case <-time.After(b.maxSeedWait):
+			b.log.Warn("Badgerdb still restoring after %s, proceeding...", b.maxSeedWait)
+		}
+	}
+
+	go func() {
+		ticker := time.NewTicker(5 * time.Minute)
+		defer ticker.Stop()
+		for {
+			select {
+			case <-b.closed:
+				return
+			case <-ticker.C:
+			}
+		again: // see https://dgraph.io/docs/badger/get-started/#garbage-collection
+			err := b.db.RunValueLogGC(0.7)
+			if err == nil {
+				goto again
+			}
+		}
+	}()
+	return nil
+}
+
+// Stop stops the repository
+func (b *Repository) Stop() error {
+	var err error
+	b.closeOnce.Do(func() {
+		close(b.closed)
+		err = b.db.Close()
+	})
+	return err
+}
+
+// Backup writes a backup of the repository to the given writer
+func (b *Repository) Backup(w io.Writer) error {
+	b.restoringLock.RLock()
+	defer b.restoringLock.RUnlock()
+	if b.restoring {
+		return model.ErrRestoring
+	}
+	if b.db.IsClosed() {
+		return badger.ErrDBClosed
+	}
+	_, err := b.db.Backup(w, 0)
+	return err
+}
+
+// Restore restores the repository from the given reader
+func (b *Repository) Restore(r io.Reader) (err error) {
+	if b.isRestoring() {
+		return model.ErrRestoring
+	}
+	if b.db.IsClosed() {
+		return badger.ErrDBClosed
+	}
+	b.setRestoring(true)
+	defer b.setRestoring(false)
+	defer func() {
+		if r := recover(); r != nil {
+			err = fmt.Errorf("panic during restore: %v", r)
+		}
+	}()
+	return b.db.Load(r, b.maxGoroutines)
+}
+
+func (b *Repository) setRestoring(restoring bool) {
+	b.restoringLock.Lock()
+	b.restoring = restoring
+	b.restoringLock.Unlock()
+}
+
+func (b *Repository) isRestoring() bool {
+	b.restoringLock.RLock()
+	defer b.restoringLock.RUnlock()
+	return b.restoring
+}
```

Comment:
we don't seems to be using this to get value of `b.restoring` at many places in this packages. Is that intentional?

Reply:
yes it is intentional, in other places we don't want to release the lock until the whole operation finishes